pkg/executor/join/hash_join_v2.go

```diff
@@ -677,166 +629,158 @@ func (e *HashJoinV2Exec) createTasks(buildTaskCh chan<- *buildTask, totalSegment
 				startIndex := partitionStartIndex[partIdx]
 				endIndex := min(startIndex+segStep, partitionSegmentLength[partIdx])
 				select {
-				case <-doneCh:
-					return
+				case <-syncer:
+					// syncer is closed by build fetcher in advance,
+					// this means that there happen some errors.
+					return false, nil
 				case buildTaskCh <- createBuildTask(partIdx, startIndex, endIndex):
 				}
 				partitionStartIndex[partIdx] = endIndex
 				hasNewTask = true
 			}
+			continue
 		}
 		if !hasNewTask {
 			break
 		}
 	}
+
+	return true, nil
 }
 
-func (e *HashJoinV2Exec) fetchAndBuildHashTable(ctx context.Context) {
-	if e.stats != nil {
-		start := time.Now()
-		defer func() {
-			e.stats.fetchAndBuildHashTable = time.Since(start)
-		}()
-	}
+func (e *HashJoinV2Exec) dispatchBuildTasks(syncer chan struct{}) error {
+	defer close(e.buildFinished)
+	ifContinue := true
+	var err error
 
-	waitJobDone := func(wg *sync.WaitGroup, errCh chan error) bool {
-		wg.Wait()
-		close(errCh)
-		if err := <-errCh; err != nil {
-			e.buildFinished <- err
-			return false
+	for ifContinue {
+		ifContinue, err = e.dispatchBuildTasksImpl(syncer)
+		if err != nil {
+			return err
+		}
+
+		// Wake up probe fetcher, we do not pass error by buildFinished in hash join v2
+		e.buildFinished <- nil
+
+		if e.stats != nil {
+			e.stats.fetchAndBuildHashTable = time.Since(e.stats.fetchAndBuildStartTime)
 		}
-		return true
 	}
+	return nil
+}
 
-	wg := new(sync.WaitGroup)
-	errCh := make(chan error, 1+e.Concurrency)
-	// doneCh is used by the consumer(splitAndAppendToRowTable) to info the producer(fetchBuildSideRows) that the consumer meet error and stop consume data
-	doneCh := make(chan struct{}, e.Concurrency)
-	srcChkCh := e.fetchBuildSideRows(ctx, wg, errCh, doneCh)
-	e.splitAndAppendToRowTable(srcChkCh, wg, errCh, doneCh)
-	success := waitJobDone(wg, errCh)
-	if !success {
-		return
+func (e *HashJoinV2Exec) startBuildFetcher(ctx context.Context) {
+	if e.stats != nil {
+		e.stats.fetchAndBuildStartTime = time.Now()
 	}
 
-	totalSegmentCnt := e.hashTableContext.mergeRowTablesToHashTable(e.hashTableMeta, e.partitionNumber)
+	// Build fetcher wakes up build workers with this channel
+	buildFetcherAndDispatcherSyncChan := make(chan struct{}, 1)
 
-	wg = new(sync.WaitGroup)
-	errCh = make(chan error, 1+e.Concurrency)
-	// doneCh is used by the consumer(buildHashTable) to info the producer(createBuildTasks) that the consumer meet error and stop consume data
-	doneCh = make(chan struct{}, e.Concurrency)
-	buildTaskCh := e.createBuildTasks(totalSegmentCnt, wg, errCh, doneCh)
-	e.buildHashTable(buildTaskCh, wg, errCh, doneCh)
-	waitJobDone(wg, errCh)
-}
+	fetcherAndWorkerSyncer := &sync.WaitGroup{}
+
+	defer func() {
+		// We must ensure that all prebuild workers have exited before
+		// set `finished` flag and close buildFetcherAndDispatcherSyncChan
+		fetcherAndWorkerSyncer.Wait()
+		e.finished.Store(true)
```

Comment:
It means this function need to wait all the probe finishes?

Reply:
> It means this function need to wait all the probe finishes?

No, build fetcher could directly exit without waiting all probe finished.